### DISCUSSION TITLE
How to check session in sunbird-ed?

### DISCUSSION THREAD

**Discussion by nidhikachhawa24** [Sunbird-Ed/Community#429] [General Discussion]
Hello,

I need to know some details about the sessions and session ID:

1. When are the sessions created for portal and mobile app?
2. What is the duration of session expire on portal?
3. Does a new sessionId get created everyday for the same user?
4. Can we record how many sessionId were generated according to days/month?

@vinukumar-vs @swayangjit @rajeevsathish

**Comment by swayangjit** [comment]
@nidhikachhawa24 I will answer this question from mobile app point of view. Sessions are recorded through telemetry events(START and END).  Mobile app session is not equivalent to portal session so session can be considered as workflows For example app start to app end can be one work flow.
1. For the guest users, every time a user opens the app a session is created with a unique sessionID and it ends once the user closes the app. For logged in user, after login a session is created with a unique sessionID and it ends when the user logs out.
3. Yes new sessionID is created when user launches the app.
4. Yes you can track the sessionID through synced telemetry.

@rajeevsathish Please provide information on sesssion in portal.

**Reply by parthlawate** [reply]
Since Android doesn't kill the app when you switch apps, do we reliably know when a session will end ?

**Reply by swayangjit** [reply]
@parthlawate As android doesn't kill the app, the session is still active. In some cases, OS kills the process when it needs more memory to process so when you switch back to the app, the app will be restarted, the previous session will be ended, and a new one will start.

**Comment by rajeshkumaravel** [comment]
@nidhikachhawa24

**From Sunbird Portal perspective**

_1. When are the sessions created for portal and mobile app?_

- **For Anonymous User(s)**
  - When user tries to access portal and if `KONG_DEVICE_REGISTER_ANONYMOUS_TOKEN` environment variable is enabled in portal backend, a session is created using node library called `express-session` along with Kong token

- **For logged in user(s)**
  - When user tries to access protected routes, post successful keycloak authentication the previous session is updated accordingly with new token from Kong

 **References** 
 `https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-5.1.0/src/app/server.js#L87`
  `https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-5.1.0/src/app/server.js#L99`

_2. What is the duration of session expire on portal?_

- **For Anonymous User(s)**
  - Anonymous user session TTL depends on environment variable `sunbird_anonymous_session_ttl`

- **For logged in user(s)**
  -  For logged in users TTL depends on environment variable `sunbird_session_ttl`

_3. Does a new sessionId get created everyday for the same user?_
  - Until user is idle for a specific period of time as mentioned with TTL duration

_4. Can we record how many sessionId were generated according to days/month?_
  - As of today we do not have explicit provision from Sunbird Portal. But the same can be achieved using telemetry data.

---
CC: @Krishnaj20 @vinukumar-vs @rajeevsathish @swayangjit

**Reply by Krishnaj20** [reply]
@chitranshu-keshav @swayangjit @rajeshkumaravel - is this thread answered, if so please mark this as locked.

**Reply by parthlawate** [reply]
@rajeshkumaravel What are the defaults for these variables ?

**Reply by rajeshkumaravel** [reply]
@parthlawate 
Please find links for default values

https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-5.2.0/src/app/helpers/environmentVariablesHelper.js#L253
https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-5.2.0/src/app/helpers/environmentVariablesHelper.js#L255
https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-5.2.0/src/app/helpers/environmentVariablesHelper.js#L65

> _Please note the above links are referred for release version **5.2.0**_

---
CC: @Krishnaj20 @vinukumar-vs @rajeevsathish @swayangjit

**Reply by parthlawate** [reply]
Thanks @rajeshkumaravel ! @nidhikachhawa24 pls note
